server/routerlicious/packages/services-shared/src/healthCheckEndpoints.ts

```diff
@@ -0,0 +1,129 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { RequestHandler, Router, Request, Response, NextFunction } from "express";
+import {
+	throttle,
+	IThrottleMiddlewareOptions,
+	getParam,
+} from "@fluidframework/server-services-utils";
+import * as core from "@fluidframework/server-services-core";
+import { StartupChecker } from "./startupChecker";
+import { LumberEventName, Lumberjack } from "@fluidframework/server-services-telemetry";
+
+export interface IReadinessCheck {
+	isReady(): Promise<boolean>;
+}
+
+export interface IThrottlerConfig {
+	tenantThrottlers: Map<string, core.IThrottler>;
+	restThrottleIdSuffix: string;
+	generalRestCallThrottleIdPrefix: string;
+}
+
+function noopMiddleware(req: Request, res: Response, next: NextFunction) {
+	next();
+}
+
+/**
+ * Creates the health check endpoints for the service.
+ * @param serviceName - The name of the service.
+ * @param readinessCheck - Optional readiness check.
+ * @param createLivenessEndpoint - Whether to create the liveness endpoint. Services like Alfred already have a ping endpoint, so this can be set to false.
+ * @param throttlerConfig - Optional throttler configuration.
+ */
+export function createHealthCheckEndpoints(
+	serviceName: string,
+	readinessCheck?: IReadinessCheck,
+	createLivenessEndpoint = true,
+	throttlerConfig?: IThrottlerConfig,
+): Router {
+	const router: Router = Router();
+	let tenantThrottleOptions: Partial<IThrottleMiddlewareOptions>;
+	let generalTenantThrottler: core.IThrottler;
+	let startupThrottler: RequestHandler;
+	let pingThrottler: RequestHandler;
+	let readinessThrottler: RequestHandler;
+
+	if (throttlerConfig) {
```

Comment:
Do we need throttlers for probes? These probes are sent at a fixed interval within the cluster from Kubernetes and there is no likelihood of a sudden surge or increase in traffic due to these probes.

Reply:
I added the throttlers to stick with the current design for ping endpoints in Alfred. Those ping endpoints do have throttling enabled.

As this would add two more endpoints which Alfred would expose - `/startup` and `/ready`, I kept the throttlers as part of the endpoints. However, you are right these are not going to see surge in traffic. So I can remove them.

Reply:
I'm wondering if we should apply this to the publicly accessible endpoints. My assumption is that the new Alfred HTTP endpoints will be public, while Riddler and Nexus will not. If that's the case, it might be beneficial to add throttlers for Alfred to mitigate potential DDoS attacks?

Reply:
Yup that was why I added them. Because for Alfred these endpoints can be pinged from any client. But these are not "well-known" endpoints. But I think for safety against DDoS we can have them. @manishgargd what do you say?

Reply:
IMO, these endpoints are so lightweight, that it would actually be counterproductive to do additional throttling checks, which rely on Redis and do more logic than these endpoints do.

It feels like overkill to throttle here.

Reply:
While `/ping` and `/startup` endpoints are quite light. `/ready` might end up being slightly heavier based on how we implement it. So I think maybe we can remove throttling from `/startup` and `/ping` but keep it for `/ready`.

Reply:
Yes, we can avoid throttlers for /ping and /startup. For /ready, can we prevent nginx to route traffic to it for public services like alfred, nexus, historian, echo. For other services, the /ready endpoint should not be available publicly.

Another option is to run /startup, /ready at ports different from 80, 443. We allow traffic only from ports 80, 443 in our clusters. Hence these endpoints will not be available publicly.

Reply:
Testing ingress of requests to the `/healthz/*` endpoints now. Will update the thread after the experiment.

Reply:
I tested and we can set ingress rules to prevent traffic from hitting these `/heathz` endpoints at an nginx level. Hence, I will remove the throttlers.